pkg/util/metrics_helper.go

```diff
@@ -6,744 +6,10 @@
-		}
-
-		if !hard && r.softRemoveUserRegistry(&r.regs[idx]) {
-			idx++ // keep it
-		} else {
-			r.regs = append(r.regs[:idx], r.regs[idx+1:]...) // remove it.
-		}
-	}
-}
-
-// Returns true, if we should keep latest metrics. Returns false if we failed to gather latest metrics,
-// and this can be removed completely.
-func (r *UserRegistries) softRemoveUserRegistry(ur *UserRegistry) bool {
-	last, err := ur.reg.Gather()
-	if err != nil {
-		level.Warn(util_log.Logger).Log("msg", "failed to gather metrics from registry", "user", ur.user, "err", err)
-		return false
-	}
-
-	for ix := 0; ix < len(last); {
-		// Only keep metrics for which we don't want to go down, since that indicates reset (counter, summary, histogram).
-		switch last[ix].GetType() {
-		case dto.MetricType_COUNTER, dto.MetricType_SUMMARY, dto.MetricType_HISTOGRAM:
-			ix++
-		default:
-			// Remove gauges and unknowns.
-			last = append(last[:ix], last[ix+1:]...)
-		}
-	}
-
-	// No metrics left.
-	if len(last) == 0 {
-		return false
-	}
-
-	ur.lastGather, err = NewMetricFamilyMap(last)
-	if err != nil {
-		level.Warn(util_log.Logger).Log("msg", "failed to gather metrics from registry", "user", ur.user, "err", err)
-		return false
-	}
-
-	ur.user = ""
-	ur.reg = nil
-	return true
-}
-
-// Registries returns a copy of the user registries list.
-func (r *UserRegistries) Registries() []UserRegistry {
-	r.regsMu.Lock()
-	defer r.regsMu.Unlock()
-
-	out := make([]UserRegistry, 0, len(r.regs))
-	out = append(out, r.regs...)
-
-	return out
-}
-
-// GetRegistryForUser returns currently active registry for given user, or nil, if there is no such registry.
-func (r *UserRegistries) GetRegistryForUser(user string) *prometheus.Registry {
-	r.regsMu.Lock()
-	defer r.regsMu.Unlock()
-
-	for idx := 0; idx < len(r.regs); idx++ {
-		if user != r.regs[idx].user {
-			continue
-		}
-
-		return r.regs[idx].reg
-	}
-	return nil
-}
-
-func (r *UserRegistries) BuildMetricFamiliesPerUser() MetricFamiliesPerUser {
-	data := MetricFamiliesPerUser{}
-	for _, entry := range r.Registries() {
-		// Set for removed users.
-		if entry.reg == nil {
-			if entry.lastGather != nil {
-				data = append(data, struct {
-					user    string
-					metrics MetricFamilyMap
-				}{user: "", metrics: entry.lastGather})
-			}
-
-			continue
-		}
-
-		m, err := entry.reg.Gather()
-		if err == nil {
-			var mfm MetricFamilyMap // := would shadow err from outer block, and single err check will not work
-			mfm, err = NewMetricFamilyMap(m)
-			if err == nil {
-				data = append(data, struct {
-					user    string
-					metrics MetricFamilyMap
-				}{
-					user:    entry.user,
-					metrics: mfm,
-				})
-			}
-		}
-
-		if err != nil {
-			level.Warn(util_log.Logger).Log("msg", "failed to gather metrics from registry", "user", entry.user, "err", err)
-			continue
-		}
-	}
-	return data
-}
-
 // FromLabelPairsToLabels converts dto.LabelPair into labels.Labels.
 func FromLabelPairsToLabels(pairs []*dto.LabelPair) labels.Labels {
```

Comment:
There's only 1 function left on this file: `GetSumOfHistogramSampleCount` it is only used in the `ha_tracker_tests.go` should we move it there?

- If yes - I don't think we need a set of tests for it as we don't usually unit tests test helpers (I think?)
- If no, then I'm happy to create a test to make sure it has some coverage.

Reply:
That makes sense to me: if the only use is in a test, let's just move it there. Testing the test helpers, I'm less sure about. If this isn't something that has changed recently (or ever), I think it's fine to drop the tests .

Reply:
Agree to leaving it and not testing it - does not seem valuable in this case.